File: src/test/matrix/py/elicitation.py
#!/usr/bin/env python3
"""MCP elicitation matrix cell server (feature=elicitation, language=py, transport=stdio).

Feature under test: SERVER-INITIATED ELICITATION.
A tool handler turns around and asks the CLIENT for structured input via the
elicitation/create server->client request (ctx.elicit(message=..., schema=...)).

Exposes:
  - 1 elicitation tool:  book_table   (handler calls ctx.elicit(...))
  - 1 plain tool:        echo         (no elicitation; keeps a trivial tool available)

Notes on expected behavior against mcpgaze:
  mcpgaze acts as a MINIMAL client during snapshot/conform/health: it advertises
  no elicitation capability and does NOT answer elicitation/create requests.
  Therefore initialize + tools/list MUST still succeed (snapshot lists 2 tools,
  conform --all PASSES, health --once is UP). Calling `book_table` will block on a
  client response that never comes, so that tools/call is expected to err/time
  out at the server side -- the key assertion is mcpgaze must NOT crash.

Built on the real Python MCP SDK (FastMCP). Elicitation schemas must be
Pydantic models with primitive field types only.
"""

from mcp.server.fastmcp import Context, FastMCP
from pydantic import BaseModel, Field

mcp = FastMCP("mcpgaze-elicitation-py")


# --- Elicitation response schema (primitive fields only) -------------------
class BookingPreferences(BaseModel):
    """Structured input requested from the client/user."""

    party_size: int = Field(description="Number of guests for the reservation.")
    confirm: bool = Field(description="Whether to confirm the booking.")
    note: str = Field(default="", description="Optional note for the restaurant.")


# --- Elicitation tool: handler asks the CLIENT for structured input --------
@mcp.tool(
    name="book_table",
    title="Book a restaurant table",
    description=(
        "Book a table by asking the connected MCP client for structured "
        "booking details. This triggers a server-initiated elicitation/create "
        "request and returns a confirmation string."
    ),
)
async def book_table(restaurant: str, ctx: Context) -> str:
    """Server-initiated elicitation: ask the client for booking preferences."""
    result = await ctx.elicit(
        message=f"Please provide booking details for {restaurant}.",
        schema=BookingPreferences,
    )
    if result.action == "accept" and result.data is not None:
        data = result.data
        if data.confirm:
            return (
                f"Booked {restaurant} for {data.party_size} "
                f"(note: {data.note or 'none'})."
            )
        return f"Booking for {restaurant} not confirmed by the user."
    return f"Booking for {restaurant} was {result.action}ed."


# --- Plain tool (no elicitation) ------------------------------------------
@mcp.tool()
def echo(message: str) -> str:
    """Echo back the provided message (no elicitation involved)."""
    return f"echo: {message}"


if __name__ == "__main__":
    mcp.run(transport="stdio")

File: src/test/matrix/ts/elicitation.mjs
#!/usr/bin/env node
// elicitation-ts matrix cell: MCP server (stdio) whose tool handler issues a
// server->client elicitation/create request (server-initiated elicitation).
// Uses the real @modelcontextprotocol/sdk. Initialize + tools/list must always
// succeed; the elicitation tool only resolves if the CLIENT answers the
// elicitation/create request with an ElicitResult.
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { ElicitResultSchema } from "@modelcontextprotocol/sdk/types.js";
import { z } from "zod";

const server = new McpServer(
  { name: "elicitation-ts", version: "1.0.0" },
  { capabilities: { tools: {} } }
);

// JSON Schema describing the structured input we want from the user. This is
// sent verbatim in the elicitation/create request's requestedSchema.
const requestedSchema = {
  type: "object",
  properties: {
    name: { type: "string", title: "Your name" },
    confirm: {
      type: "boolean",
      title: "Proceed?",
      description: "Whether to proceed with the action",
    },
  },
  required: ["confirm"],
};

// The elicitation tool: its handler asks the CLIENT for structured input via a
// server->client `elicitation/create` request. We issue the request at the
// lowest level (server.server.request) so it is actually put on the wire
// regardless of whether the client advertised an elicitation capability -- the
// public elicitInput() helper short-circuits with a thrown error when the
// client capability is absent, which would never exercise the wire. If the
// client never answers, we bound the wait with the request's own timeout so the
// tools/call resolves to a clean error instead of hanging forever.
server.registerTool(
  "ask_user",
  {
    title: "Ask the user for structured input",
    description:
      "Issue a server-initiated elicitation/create request asking the " +
      "connected client to collect structured input (name + confirm), then " +
      "return the user's response.",
    inputSchema: {
      prompt: z
        .string()
        .optional()
        .describe("Optional custom prompt message shown to the user"),
    },
  },
  async ({ prompt }) => {
    const message =
      prompt ?? "Please provide your name and confirm whether to proceed.";

    // Low-level server->client request: method "elicitation/create". Bounded so
    // a non-answering client yields an error, not an indefinite hang.
    const result = await server.server.request(
      {
        method: "elicitation/create",
        params: { message, requestedSchema },
      },
      ElicitResultSchema,
      { timeout: 8000 }
    );

    if (result.action !== "accept") {
      return {
        content: [
          { type: "text", text: `user did not accept (action=${result.action})` },
        ],
      };
    }

    return {
      content: [
        {
          type: "text",
          text: `user provided: ${JSON.stringify(result.content ?? {})}`,
        },
      ],
    };
  }
);

// A plain tool so tools/list works and is meaningful even without elicitation.
server.registerTool(
  "echo",
  {
    title: "Echo",
    description: "Echo back the provided message (no elicitation involved)",
    inputSchema: { message: z.string().describe("Message to echo back") },
  },
  async ({ message }) => ({
    content: [{ type: "text", text: `echo: ${message}` }],
  })
);

const transport = new StdioServerTransport();
await server.connect(transport);
